src/tools/build_helper/src/git.rs

```diff
@@ -167,29 +167,59 @@ pub fn get_git_untracked_files(
 ///
 /// This can result in formatting thousands of files instead of a dozen,
 /// so we should warn the user something is wrong.
-pub fn warn_old_master_branch(
-    config: &GitConfig<'_>,
-    git_dir: &Path,
-) -> Result<(), Box<dyn std::error::Error>> {
-    use std::time::Duration;
-    const WARN_AFTER: Duration = Duration::from_secs(60 * 60 * 24 * 10);
-    let updated_master = updated_master_branch(config, Some(git_dir))?;
-    let branch_path = git_dir.join(".git/refs/remotes").join(&updated_master);
-    match std::fs::metadata(branch_path) {
-        Ok(meta) => {
-            if meta.modified()?.elapsed()? > WARN_AFTER {
-                eprintln!("warning: {updated_master} has not been updated in 10 days");
-            } else {
-                return Ok(());
+pub fn warn_old_master_branch(config: &GitConfig<'_>, git_dir: &Path) {
+    if crate::ci::CiEnv::is_ci() {
+        // this warning is useless in CI,
+        // and CI probably won't have the right branches anyway.
+        return;
+    }
+    // this will be overwritten by the actual name, if possible
+    let mut updated_master = "the upstream master branch".to_string();
+    match warn_old_master_branch_(config, git_dir, &mut updated_master) {
+        Ok(branch_is_old) => {
+            if !branch_is_old {
+                return;
             }
+            // otherwise fall through and print the rest of the warning
         }
         Err(err) => {
             eprintln!("warning: unable to check if {updated_master} is old due to error: {err}")
         }
     }
     eprintln!(
         "warning: {updated_master} is used to determine if files have been modified\n\
-               warning: if it is not updated, this may cause files to be needlessly reformatted"
+         warning: if it is not updated, this may cause files to be needlessly reformatted"
     );
-    Ok(())
+}
+
+pub fn warn_old_master_branch_(
+    config: &GitConfig<'_>,
+    git_dir: &Path,
+    updated_master: &mut String,
+) -> Result<bool, Box<dyn std::error::Error>> {
+    use std::time::Duration;
+    *updated_master = updated_master_branch(config, Some(git_dir))?;
+    let branch_path = git_dir.join(".git/refs/remotes").join(&updated_master);
```

Comment:
Isn't there a way to avoid directly looking at git's data format? Breaking the abstraction like this is not a good idea.

If you must use the .git folder, please use `git rev-parse --git-dir` to determine its location, rather than assuming that it is `.git`. That will then also work with worktrees.

Reply:
Ah, I see you already found `--git-common-dir`... which is indeed the right location, `--git-dir` is wrong. Let me quickly fix some faulty logic I recently wrote elsewhere, then... ;)

Reply:
i suppose i overestimated:
1. how many people use worktrees
2. how much people dislike incorrect warnings

waiting on a second opinion, i'll rewrite the code if that's the consensus preference.

Reply:
The incorrect warning already led to incorrect diagnosis of an unrelated issue in https://github.com/rust-lang/rust/issues/130144. So it's not just about dislike, it's about the real costs of an incorrect warning.

Reply:
yeah, i see that pretty clearly now, there's a reason i started writing this PR as soon as i was made aware of the issue.

i know the current PR works, and my main worry is that if i need to totally rewrite it, that might introduce even more bugs (this situation has happened to me many times), and this would be even worse with something that is so tedious for me to test (spinning up worktrees is not fast on a hdd system)

Reply:
> i started writing this PR as soon as i was made aware of the issue.

Thanks for spinning up a fix so quickly!